Add `keg_only :versioned_formula`.

MikeMcQuaid opened the pull request:
This is used to indicate a formula is a version of another formula. This will be used to provide a consistent interface for older formulae versions and replaces the use of `conflicts_with`.

CC @ilovezfs as we talked about this.

MikeMcQuaid commented:
Heh, second duplicate of https://github.com/Homebrew/brew/pull/1844 because I keep pushing the wrong branch.

MikeMcQuaid commented:
Not a duplicate PR actually, just duplicate title/body (which I've edited).

ilovezfs commented on docs/Versions.md:
```diff
@@ -11,6 +11,6 @@ Versioned formulae we include must meet the following standards:
 * Versioned formulae should differ in major/minor (not patch) versions from the current stable release. This is because patch versions indicate bug or security updates and we want to ensure you apply security updates.
 * Formulae that depend on versioned formulae must not depend on the same formulae at two different versions twice in their recursive dependencies. For example, if you depend on `openssl@1.0` and `foo`, and `foo` depends on `openssl` then you must instead use `openssl`.
-* Versioned formulae should strive to be linked at the same time as their non-versioned counterpart (without patching). If this is not possible, favour either `conflicts_with` or `keg_only` depending on whether users expect to have multiple versions installed at once or not.
+* Versioned formulae should only be linkable at the same time as their non-versioned counterpart if the upstream project provides support for e.g. suffixed binaries. If this is not possible, use `keg_only :version_formula` depending to allow users to have multiple versions installed at once.
```
"depending to allow" -> "to allow"

ilovezfs replied:
Not sure we should allow any suffixing even if it's available.

ilovezfs commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -459,6 +459,14 @@ def audit_deps
   end
 
   def audit_conflicts
+    if formula.name.include?("@")
```
might want to put this in `formula.rb` behind a method like `is_version_formula?` to avoid hard coding the "@" character as meaning versioned and hard coding versioned as must have "@".

ilovezfs commented on Library/Homebrew/formula_support.rb:
```diff
@@ -29,27 +29,32 @@ def valid?
-      "Xcode provides this software prior to version 4.3."
-    when :provided_until_xcode5
-      "Xcode provides this software prior to version 5."
+    when :version_formula then <<-EOS.undent
```
You don't like `:versioned_formula`?

MikeMcQuaid commented:
@ilovezfs Addressed all comments except have allowed suffixing for now. It seems to be useful at least for GCC and anything else that has a full path specified like that.

MikeMcQuaid commented:
Had to rerun CI due to a flaky Cask test.
